Upgrade and Expansion of Detailed Ice Storage Performance Curves

RKStrand opened the pull request:
Pull request overview
---------------------
Input fields related to curve fits were not in line with other input syntax that was recently changed.  So, this was addressed by this defect.  The capabilities were expanded in two ways.  First, now any two independent variable curve can be used.  In addition, there are some new options for what those independent variables are thus further expanding the ability to more accurately model detailed ice storage systems in EnergyPlus.

Also, corrected use of the `MassFlowstar` variable that was not necessarily initialized before being used. This changes results for simulations with ThermalStorage:Ice:Detailed using the new LMTDMassFlow curve variable specification (previously the CubicLinear Curve Object Type).

### Work Checklist
Add to this list or remove from it as applicable.  This is a simple templated set of guidelines.
 - [x] Title of PR should be user-synopsis style (clearly understandable in a standalone changelog context)
 - [x] At least one of the following appropriate labels must be added to this PR to be consumed into the changelog:
   - Defect: This pull request repairs a github defect issue #6879 

### Review Checklist
This will not be exhaustively relevant to every PR.
 - [x] Functional code review (it has to work!)
 - [x] If defect, results of running current develop vs this branch should exhibit the fix
 - [x] CI status: all green or justified
 - [x] Performance: CI Linux results include performance check
 - [x] Unit Test(s)
 - C++ checks:
   - [x] Argument types
   - [x] If any virtual classes, ensure virtual destructor included, other things
 - IDD changes:
   - [x] Verify naming conventions and styles, memos and notes and defaults
   - [x] Open windows IDF Editor with modified IDD to check for errors
   - [x] If transition, add to input rules file for interfaces
   - [x] If transition, add transition source
   - [x] If transition, update idfs
 n/a If new idf included, locally check the err file and other outputs
 - [x] Required documentation updates?
 n/a ExpandObjects changes?
 n/a If output changes, including tabular output structure, add to output rules file for interfaces

RKStrand committed:
Initial solution commit

First pass at changes to address this issue.  Includes changes to the
IDD and the two input files that use the detailed ice storage model in
the test suite.

RKStrand committed:
Merge branch 'develop' into 6879-ChangeDetIceStorCurveTypeFields

RKStrand committed:
Merge branch 'develop' into 6879-ChangeDetIceStorCurveTypeFields

RKStrand committed:
Minor fixes to error messages

Small changes to error messages to reflect what is being done

RKStrand committed:
Changes to integration input files

Changes to integration input files in the test suite

RKStrand committed:
Fixed a bug/typo in the solution

A bug was discovered in the solution that involved a cut and paste that
wasn’t fixed properly.

RKStrand committed:
Unit test and documentation

Final changes for this bug including the unit test and documentation
corrections based on this fix.

RKStrand committed:
Transition change for IDD

This is the required change to any input files from previous editions
of the IDD.

RKStrand commented:
@mjwitte Would you like to review this since you were part of the discussion on the defect description?  If so, please change the Pivotal item to reflect that you are reviewing it.  Also, you will have one output file that will show differences.  That is because another error was uncovered during work on this mod where develop was using a variable that was not necessarily initialized before being used.  It was a variable that determined the performance of the system (MassFlowstar).  This is now fixed so differences are to be expected.

mjwitte committed:
Merge remote-tracking branch 'remotes/origin/develop' into 6879-ChangeDetIceStorCurveTypeFields

mjwitte committed:
Add transition for ice storage

mjwitte commented on idd/Energy+.idd.in:
```diff
@@ -76268,17 +76268,21 @@ ThermalStorage:Ice:Detailed,
        \required-field
        \type node
-  A5 , \field Discharging Curve Object Type
+  A5 , \field Discharging Curve Variable Specifications
```
This field and "Charging Curve Variable Specifications" should have a default, or be required.
Should the curve names be required?
Any other fields that should be required or have a default?

RKStrand replied:
@mjwitte Made all four required fields.  No defaults--the user has to make some sort of decision here.

mjwitte commented on doc/engineering-reference/src/simulation-models-encyclopedic-reference-001/ice-thermal-storage.tex:
```diff
@@ -103,7 +103,11 @@ \subsection{Detailed Ice Storage Model}\label{detailed-ice-storage-model}
-The actual performance of the ice storage unit depends on the physical geometry, materials, and characteristics of the ice storage unit.~ In analyzing performance data trends from ice storage manufacturers, it was determined that the following equation would work well for the discharging process:
+The actual performance of the ice storage unit depends on the physical geometry, materials, and characteristics of the ice storage unit. Modeling both the charging and discharging performance of a particular ice storage unit is accomplished in EnergyPlus using Curve objects that establish the relationships between various parameters of the unit and the output of the unit itself. The input that co ...
+
+The Variable Specifications input fields determine which parameters are used as the independent variables in the curves referenced by the Curve Name inputs.  There are four options for the Variable Specifications fields: PercentChargedLMTD, PercentDischargedLMTD, LMTDMassFlow, and LMTDPercentCharged.  Each option defines which two variables are the independent variables in the curve equation as w ...
```
Looks like you changed your mind from PercentCharged/Discharged* to FractionCharged/Discharged? The IDD and code use Fraction, but the docs are mostly Percent.

RKStrand replied:
@mjwitte Good catch.  My bad.  I have fixed this so there are no more Percent*** in this doc.

mjwitte commented on doc/input-output-reference/src/overview/group-plant-equipment.tex:
```diff
@@ -6275,21 +6275,21 @@ \subsubsection{Inputs}\label{inputs-21-004}
+\paragraph{Field: Discharging Curve Variable Specifications}\label{field-discharging-curve-variable-specifications}
 
-The detailed ice storage model in EnergyPlus takes advantage of the Curve feature of the program. Currently, the only two allowed curve fit types for the detailed ice storage model are the QuadraticLinear and the CubicLinear curves. More information on this curve can be found in the section on Curves.
+The detailed ice storage model in EnergyPlus takes advantage of the Curve feature of the program.  The discharging curve can use any curve that has two independent variables.  This field controls which two parameters are the independent variables in modeling the performance of the detailed ice storage model during discharging. Currently, there are four different independent variable specification ...
```
This would read better as an itemized list. 
And (nit-picking) no need to say "Currently, there are." --> "There are four . . ."

RKStrand replied:
@mjwitte Fixed this...and correct the Percent typos here as well.

mjwitte commented on doc/input-output-reference/src/overview/group-plant-equipment.tex:
```diff
@@ -6275,21 +6275,21 @@ \subsubsection{Inputs}\label{inputs-21-004}
+\paragraph{Field: Charging Curve Variable Specifications}\label{field-charging-curve-variable-specifications}
 
-The detailed ice storage model in EnergyPlus takes advantage of the Curve feature of the program. Currently, the only two allowed curve fit types for the detailed ice storage model are the QuadraticLinear and the CubicLinear curves. More information on this curve can be found in the section on Curves.
+The detailed ice storage model in EnergyPlus takes advantage of the Curve feature of the program.  The charging curve can use any curve that has two independent variables.  This field controls which two parameters are the independent variables in modeling the performance of the detailed ice storage model during charging. Currently, there are four different independent variable specifications: Per ...
```
Same here, itemized list.

RKStrand replied:
@mjwitte Yup...got it here as above.

mjwitte commented on testfiles/5ZoneDetailedIceStorageCubicLinear.idf:
```diff
@@ -3058,9 +3058,9 @@
+    LMTDMassFlow,            !- Discharging Curve Object Type
     DischargeCurve,          !- Discharging Curve Name
-    CubicLinear,             !- Charging Curve Object Type
+    LMTDMassFlow,            !- Charging Curve Object Type
```
Need new field names here and in some of the other idfs.

RKStrand replied:
@mjwitte I think I got all of them now.  Two were already done.  Missed the others in the previous pass.  Good catch, again.

mjwitte commented on tst/EnergyPlus/unit/IceThermalStorage.unit.cc:
```diff
@@ -0,0 +1,180 @@
+    {
+        EnergyPlusFixture::TearDown(); // Remember to tear down the base fixture after cleaning up derived fixture!
+    }
+};
```
Looks like you were going to make a new fixture `IceThermalStorageTest` but I don't see it used below. So this class section can be removed.

RKStrand replied:
@mjwitte A case of monkey see, monkey do.  Grabbed an existing file as a starting point and should have figured out that I didn't need that stuff.  It's all gone now.

mjwitte commented on tst/EnergyPlus/unit/IceThermalStorage.unit.cc:
```diff
@@ -0,0 +1,180 @@
+    Real64 Tolerance = 0.001;
+    
+    if (allocated(IceThermalStorage::DetIceStor)) DetIceStor.deallocate();
+    if (allocated(CurveManager::PerfCurve)) PerfCurve.deallocate();
```
These should be deallocated in a `clear_state` function. Oh, I see, IceThermalStorage doesn't have one. Follow the example of, say Pumps.cc. Every module-level variable should be reset to it's initial value and every array deallocated in a `clear_state` function, and then the new function needs to get added to [`EnergyPlusFixture::clear_all_states()`](https://github.com/NREL/EnergyPlus/blob/af8b6e160fa36a0cc5d5eaa28db46883e97a2e96/tst/EnergyPlus/unit/Fixtures/EnergyPlusFixture.cc).

RKStrand replied:
@mjwitte Ok, IceThermalStorage now has a clear_state and the checks noted above have been deleted.

mjwitte commented on src/EnergyPlus/IceThermalStorage.cc:
```diff
@@ -1109,15 +1096,20 @@ namespace IceThermalStorage {
+                DetIceStor(IceNum).DischargeCurveTypeNum = CurveVarsLMTDFracCharged;
             } else {
-                ShowSevereError(cCurrentModuleObject + ": Discharge curve type not valid, type=" + cAlphaArgs(5));
+                ShowSevereError(cCurrentModuleObject + ": Discharge curve independent variable options not valid, option=" + cAlphaArgs(5));
```
If I'm following correctly, the user could select a valid choice for variable type, but specify a curve with the wrong dimensions and this error message would fire. It would be cleaner to separate these two and have separate error messages for invalid variable options or invalid curve dimension. Same for the charge curve.

RKStrand replied:
@mjwitte Now separated, for both discharging and charging.

mjwitte commented on src/EnergyPlus/IceThermalStorage.cc:
```diff
@@ -2360,6 +2357,33 @@ namespace IceThermalStorage {
         return CalcDetIceStorLMTDstar;
     }
 
+    Real64 CalcQstar(int const CurveIndex,      // curve index
```
Nice to move this to a separate function.

RKStrand replied:
@mjwitte Thanks!  Makes unit test creation much cleaner as well.

mjwitte commented on tst/EnergyPlus/unit/IceThermalStorage.unit.cc:
```diff
@@ -0,0 +1,180 @@
+    
+    //Test 1: CurveVarsFracChargedLMTD Curve is QuadraticLinear
+    TestNum = 1;
+    IceStorageCurveType = 1; // FracChargedLMTD
```
Nice unit test. It would be more robust and readable to use the actual parameter variables to assign `IceStorageCurveType` and `PerfCurve.CurveType`.

e.g. `IceStorageCurveType = IceThermalStorage::CurveVarsFracChargedLMTD`.

RKStrand replied:
@mjwitte Took care of this as well. Will make the commits and if everything comes back as before (hopefully the case), this will be ready for re-review.

mjwitte reviewed (changes requested):
This is a nice improvement. 

I've merged in develop, and added the transition code. I tested it on all of the detailed ice storage idfs - please review and test it. BTW, I also added the RunPeriod transition rule description which was overlooked in that PR. More comments below.

mjwitte committed:
Transition for ice storage - fraction instead of percent

RKStrand committed:
Merge branch 'develop' into 6879-ChangeDetIceStorCurveTypeFields

RKStrand committed:
Addressed review comments

Various changes to address the review comments for this work.

mjwitte commented:
@RKStrand You may have noticed this build error already 
‘clear_state’ is not a member of ‘EnergyPlus::IceThermalStorage’. 

It's need a declaration in IceThermalStorage.hh

RKStrand commented:
Mike:

Yes, I saw that and have already fixed it…but for some reason I forgot to commit it to GitHub.  Will rectify that momentarily…

Thanks,

Rick


On Jul 10, 2019, at 5:15 PM, Michael J. Witte <notifications@github.com<mailto:notifications@github.com>> wrote:


@RKStrand<https://github.com/RKStrand> You may have noticed this build error already
‘clear_state’ is not a member of ‘EnergyPlus::IceThermalStorage’.

It's need a declaration in IceThermalStorage.hh

—
You are receiving this because you were mentioned.
Reply to this email directly, view it on GitHub<https://github.com/NREL/EnergyPlus/pull/7339?email_source=notifications&email_token=ABMJHPNB6JXG7POMOIWZYB3P6ZNR5A5CNFSM4HYYJIX2YY3PNVWWK3TUL52HS4DFVREXG43VMVBW63LNMVXHJKTDN5WW2ZLOORPWSZGODZU5B3Q#issuecomment-510251246>, or mute the thread<https://github.com/notifications/unsubscribe-auth/ABMJHPPSW7XYRKQS6DZRCOTP6ZNR5ANCNFSM4HYYJIXQ>.

RKStrand committed:
Correct a few overlooked things

A few things were not done properly in the previous commit.  This
should all be corrected now.

RKStrand commented:
@mjwitte Ok, I think I have resolved the issues you addressed and things are back to where they were again.  So, I think this is ready for another review.

mjwitte committed:
Ice storage - fix typos and other doc cleanup

mjwitte committed:
Merge remote-tracking branch 'remotes/origin/develop' into 6879-ChangeDetIceStorCurveTypeFields

mjwitte reviewed (approved):
Unit tests run serially are all good now.
Ran all of the detailed ice storage test files as annual simulations - only a couple of warnings in 2 of the files; others all clear.
Changes to address prior comments look good.
Fixed a couple of typos and merged in develop one last time.
This will be merged once CI finishes.

mjwitte committed:
Merge remote-tracking branch 'remotes/origin/develop' into 6879-ChangeDetIceStorCurveTypeFields

# Conflicts:
#	src/Transition/InputRulesFiles/Rules9-1-0-to-9-2-0.md

mjwitte committed:
Cleanup merge conflict

jmarrec commented on doc/input-output-reference/src/overview/group-plant-equipment.tex:
```diff
@@ -6325,9 +6352,9 @@ \subsubsection{Inputs}\label{inputs-21-004}
       Ice Tank Inlet Node,     !- Plant Loop Inlet Node
       Ice Tank Outlet Node,    !- Plant Loop Outlet Node
-      QuadraticLinear,         !- Discharging Curve Fit Type
+      PercentDischargedLMTD,   !- Discharging Curve Fit Type
```
Incorrect, should be `FractionDischargedLMTD`

jmarrec commented on doc/input-output-reference/src/overview/group-plant-equipment.tex:
```diff
@@ -6325,9 +6352,9 @@ \subsubsection{Inputs}\label{inputs-21-004}
+      PercentDischargedLMTD,   !- Discharging Curve Fit Type
       DischargeCurve,          !- Discharging Curve Name
-      QuadraticLinear,         !- Charging Curve Fit Type
+      PercentChargedLMTD,      !- Charging Curve Fit Type
```
Should be FractionChargedLMTD

jmarrec reviewed (commented):
Typo in doc

RKStrand reviewed (commented):
@jmarrec Good catch!  Thanks for finding that and making the change.

@Myoldmopar These changes are correct and can be merged whenever it is convenient to do so.  It's just a documentation change so it shouldn't result in any changes to the outputs.

RKStrand commented:
@Myoldmopar I should have mentioned that @jmarrec fixed this in a separate issue #7485
